docs/ocdump.1
.TH OCDUMP 1
.SH NAME
ocdump \- Dump the DAS, DDS, and/or DATADDS for a given OPeNDAP source
.SH SYNOPSIS
.HP
ocdump
.nh
\%[-B]
\%[-C \fIconstraint-string\fP]
\%[-D \fIdebug-argument\fP]
\%[-g]
\%[-m]
\%[-A]
\%[-p \fIall|das|dds|data|datadds\fP]
\%[-U \fIuser-parameters\fP]
\%[-v]
\%[-L]
\%[-m]
\%[[-f] \fIOPeNDAP-URL\fP]
.hy
.ft
.SH DESCRIPTION
\fBoctest\fP
takes an OPeNDAP source (via a URL) and prints various kinds of information from that source.
Independent of the options, 
\fBoctest\fP
will always attempt to obtain a DAS and DDS.
Failures will be reported but ignored.
The DATADDS will only be accessed if explicitly requested using \fI-p datadds\fP
option.
.LP
.SH OPTIONS
.IP "\fB-B\fP"
Produce the output in a form using nested curly brackets;
default is off.
.IP "\fB-C \fRconstraint\fP"
Specify the constraint to be used when the DATADDS is read.
.IP "\fB-g\fP"
Force \fI-p\fP flag to generate the DAS and/or DDS
from the internal parse tree instead of just
dumping the text sent by the server.
.IP "\fB-m\fP"
Store the data from the data dds in memory instead of in a temporary
file. As always, there is speed/space tradeoff.
.IP "\fB-A\fP"
Print attributes along with the DDS if \fI-g\fP is specified.
The -M flag is an older alias for -A, but -M is deprecated.
.IP "\fB-p \fRall|das|dds|data|datadds\fP"
Cause the DAS, the DDS, or the DATADDS to be printed.
The argument \fIdata\fP is a synonym for \fIdatadds\fP.
The  \fIall\fP case causes all three to be printed.
This option may be used multiple times.
.IP "\fB-U \fRuser-parameters\fP"
Prefix the given parameters to the front of the OPeNDAP URL
before accessing it. The general form is param=value.
.IP "\fB-v\fP"
Output extra information about processing.
.IP "\fB-L\fP"
Turn on logging output.
.IP "\fB-D \fRdebug-arg\fP"
Set a variety of debug options; this flag may be repeated multiple times.
See \fIocdump.c\fP for a detailed description.
.IP "\fB-f \fROPeNDAP-URL\fP"
Specify the url source from which to read. The -f is optional.
.SH EXAMPLES
.LP
Extract everything from a specific URL (note: URL is fake).
.RS
.HP
ocdump -p all http://test.opendap.org/test1
.RE

.SH BUGS
.LP
Probably many; but currently none known.



ChangeLog
2012-07-05  dmh

	fix bug when reading empty .dodsrc file

2012-07-05  dmh

	undo mixup from moving to https

2012-07-05  dmh

	migrate to https repository

2012-06-23  dmh

	added oc_data_read variants

2012-06-23  dmh

	re-add oc-ping

2012-06-22  dmh

	added oc_ping to API

2012-06-21  dmh

	make oc_raw_xdrsize return error

2012-06-21  dmh

	fix oc_raw_xdrsize

2012-06-21  dmh

2012-06-20  dmh

	distributable version of oc2.0

2012-06-20  dmh

	checkpoint

2012-06-20  dmh

	pre-release

2012-06-20  dmh

	checkpoint

2012-06-20  dmh

	update docs/html dir

2012-06-20  dmh

	extend documentation

2012-06-20  dmh

	ckp

2012-06-19  dmh

	first distcheck version

2012-06-19  dmh

	first version to pass  all tests

2012-06-18  dmh

2012-06-18  dmh

2012-06-15  dmh

	checkpoint before removing instance

2012-06-14  dmh

2012-06-14  dmh

2012-06-13  dmh

	checkpoint

2012-06-09  dmh

2012-06-07  dmh

2012-06-06  dmh

2012-06-06  dmh

2012-06-06  dmh

2012-06-06  dmh

2012-06-05  dmh

2012-06-03  dmh

2012-06-01  dmh

2012-05-30  dmh

2012-05-18  dmh

2012-05-16  dmh

2012-05-16  dmh

2012-05-12  dmh

	Rebuild the interface

2012-05-12  dmh

2012-05-12  dmh

2012-05-11  dmh

	fork to experiment with new oc_data interface

2012-05-10  dmh

	new data operators

2012-05-08  dmh

	experimental ocpath proposal

2012-05-04  dmh

	update automake files

2012-05-04  dmh

	disconnect mode constants from object types

2012-05-04  dmh

	add some explanatory text about oc_data procedures.

2012-04-30  dmh

	ckp

2012-04-30  dmh

	cleanup from file rename

2012-04-30  dmh

	rename files to avoid conflicts

2012-04-22  dmh

	changed port of the remote test server

2012-04-20  dmh

	change IGNORE ifdef flag to OCIGNORE; apparently libcurl defines IGNORE

2012-04-05  dmh

2012-04-05  dmh

2012-04-05  dmh

	added missing docs/Makefile.am

2012-04-02  dmh

	improve documentation

2012-04-01  dmh

	prepare for new release

2012-03-21  dmh

	cleanup

2012-03-21  dmh

2012-03-16  dmh

2012-03-14  dmh

	reduce documentation load

2012-03-14  dmh

	final cleanup; make documentation generated by make dist

2012-03-14  dmh

	do not include the generated documenation in svn

2012-03-14  dmh

	cleanup and revert

2012-03-14  dmh

2012-03-14  dmh

	remove 'old' dir

2012-03-11  dmh

	improve error reporting out of curl

2012-01-28  dmh

	separate ocdebug and ocdump

2012-01-28  dmh

	checkpoint

2012-01-28  dmh

	ckp

2012-01-28  dmh

	1) install ocdump==octest; 2) allow for zero length strings in a packet

2012-01-25  dmh

	memory leak

2012-01-25  dmh

	fix memory leak

2012-01-25  dmh

	fixed memory leaks

2012-01-25  dmh

	remove swapinlinedouble

2012-01-24  dmh

	typo

2012-01-24  dmh

	modify the transition table

2012-01-23  dmh

	remove need for tempnam

2012-01-23  dmh

	remove debugging

2012-01-23  dmh

	fix xdr double conversion

2012-01-23  dmh

	add copies of the manuals

2012-01-23  dmh

	ckp

2012-01-23  dmh

	ckp

2012-01-21  dmh

	revert back to version 1.0 since interface is unchanged

2012-01-21  dmh

	update version to 1.1

2012-01-21  dmh

	fixed typo in testcase expected output

2012-01-21  dmh

	major rewrite of the oc data walking system to improve performance

2012-01-17  dmh

	ckp

2012-01-13  dmh

	ckp

2012-01-13  dmh

	ckp

2012-01-12  dmh

	ckp

2012-01-12  dmh

2012-01-11  dmh

	ckp

2012-01-10  dmh

	ckp

2012-01-07  dmh

	ckp

2012-01-07  dmh

	checkpoint

2012-01-06  dmh

	move keys to separate dir

2012-01-06  dmh

	move test keys to separate directory

2012-01-06  dmh

	working on authorization

2012-01-06  dmh

	minor tweek to testauth.sh

2012-01-05  dmh

	client side keys

2012-01-05  dmh

	auth checks

2011-12-18  dmh

2011-12-11  dmh

	abandon git; continue svn development

2011-12-10  dmh

2011-11-17  dmh

	cloned

2011-11-17  dmh

	fixed test cases

2011-08-01  dmh

	fixed Data: handling

2011-08-01  dmh

	fixed Data: handling

2011-07-28  dmh

	fix bug in ocuri.c when url has only host name

2011-07-17  dmh

	fixed escaping problem

2011-07-06  dmh

2011-07-06  dmh

2011-06-02  dmh

	make changes to help support WIN32

2011-05-25  dmh

	fix test cases to match new escape handling

2011-05-25  dmh

	fix escape handling

2011-05-24  dmh

	added OC_ENOFILE error

2011-05-13  dmh

	added timeout to .dodsrc

2011-05-12  dmh

	added some aliasing capabilities to the .dodsrc processing

2011-05-03  dmh

	handle \!HAVE_STRDUP

2011-05-03  dmh

	name conflicts

2011-05-03  dmh

	added use of CURLOPT_MAXREDIRS

2011-04-26  dmh

	parameter handling

2011-04-26  dmh

	always enable cookies

2011-04-25  dmh

	ditto

2011-04-25  dmh

	more password foolery

2011-04-20  dmh

	fixed url parsing

2011-04-20  dmh

	fixed url parsing

2011-04-20  dmh

	genericized dapurl

2011-04-20  dmh

	fixed password problem

2011-04-20  dmh

	fixed password handling

2011-04-14  dmh

	made the bisoninvocation optional

2011-04-06  dmh

	fixed some warnings

2011-03-31  dmh

	made sure config.h was first in all files

2011-03-28  dmh

	fix rare bug in data extraction

;; Local Variables:
;; coding: utf-8
;; End:
2011-01-21  dmh
	Added complete rewrite of the xdr data walking code in occontent.c

2010-07-18  dmh

	Added url prefix support to .dodsrc file.

2010-07-18  dmh

	At request of ESG, modified .dodsrc processing to support SSL.

2009-08-20  rduncan

	Added doxygen generated documentation

2009-08-13  dmh

	removed non-autotest testing

2009-08-13  James Gallagher <jgallagher@opendap.org>

	Removed the executable property on the .c and .h files

	_M   oc.c
	_M   oc.h
	_M   ocdebug.c
	_M   ocdebug.h
	_M   http.c
	_M   ocxdr_stdio.c
	_M   ocdump.c
	_M   ocdump.h
	_M   occontent.c
	_M   imain.c
	_M   occontent.h
	_M   getoc.c
	_M   ocnode.c
	_M   ocnode.h
	_M   ocdrno.c
	_M   ocdrno.h
	_M   ocutil.c
	_M   ocutil.h
	_M   occompile.c
	_M   ocdatatypes.h
	_M   ocdata.c
	_M   ocdata.h
	_M   octest.c
	_M   ocinternal.h
	_M   oclog.c
	_M   oclog.h
	_M   i.c
	_M   i.h

2009-08-04  dmh

	bug fixes


ocdump.c
/* Copyright 2009, UCAR/Unidata and OPeNDAP, Inc.
   See the COPYRIGHT file for more information.
*/

#include "config.h"
#include <sys/stat.h>

#ifdef NETINET_IN_H
#include <netinet/in.h>
#endif

#ifndef HAVE_SSIZE_T
#ifdef SSIZE_T
#define ssize_t SSIZE_T
#else
#define ssize_t int
#endif
#define HAVE_SSIZE_T
#endif 

#include "ocinternal.h"
#include "ocdebug.h"

#define MAXLEVEL 1

/*Forward*/
static void dumpocnode1(OCnode* node, int depth);
static void dumpdimensions(OCnode* node);
static void dumpattvalue(OCtype nctype, const char*);

static char* sindent = NULL;

static char*
dent(int n)
{
    if(sindent == NULL) {
	sindent = (char*)ocmalloc(102);
	MEMCHECK(sindent,NULL);
	memset((void*)sindent,(int)' ',(size_t)101);
	sindent[101] = '\0';
    }
    if(n > 100) n = 100;
    return sindent+(100-n);
}

/* support [dd] leader*/
static char*
dent2(int n) {return dent(n+4);}

static void
tabto(int pos, OCbytes* buffer)
{
    int bol,len,pad;
    len = ocbyteslength(buffer);
    /* find preceding newline */
    for(bol=len-1;;bol--) {
	int c = ocbytesget(buffer,(size_t)bol);
	if(c < 0) break;
	if(c == '\n') {bol++; break;}
    }
    len = (len - bol);
    pad = (pos - len);
    while(pad-- > 0) ocbytescat(buffer," ");
}

void
ocdumpnode(OCnode* node)
{
    if(node != NULL) {
        dumpocnode1(node,0);
    } else {
	fprintf(stdout,"<NULL>\n");
    }
    fflush(stdout);
}

static void
dumpocnode1(OCnode* node, int depth)
{
    unsigned int n;
    switch (node->octype) {
    case OC_Atomic: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	if(node->name == NULL) OCPANIC("prim without name");
	fprintf(stdout,"%s %s",octypetostring(node->basetype),node->name);
	dumpdimensions(node);
	fprintf(stdout," &%lx",(unsigned long)node);
	fprintf(stdout,"\n");
    } break;

    case OC_Dataset: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	fprintf(stdout,"dataset %s\n",
		(node->name?node->name:""));
	for(n=0;n<oclistlength(node->subnodes);n++) {
	    dumpocnode1((OCnode*)oclistget(node->subnodes,n),depth+1);
	}
    } break;

    case OC_Structure: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	fprintf(stdout,"struct %s",
		(node->name?node->name:""));
	dumpdimensions(node);
	fprintf(stdout," &%lx",(unsigned long)node);
	fprintf(stdout,"\n");
	for(n=0;n<oclistlength(node->subnodes);n++) {
	    dumpocnode1((OCnode*)oclistget(node->subnodes,n),depth+1);
	}
    } break;

    case OC_Sequence: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	fprintf(stdout,"sequence %s",
		(node->name?node->name:""));
	dumpdimensions(node);
	fprintf(stdout," &%lx",(unsigned long)node);
	fprintf(stdout,"\n");
	for(n=0;n<oclistlength(node->subnodes);n++) {
	    dumpocnode1((OCnode*)oclistget(node->subnodes,n),depth+1);
	}
    } break;

    case OC_Grid: {
	unsigned int i;
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	fprintf(stdout,"grid %s",
		(node->name?node->name:""));
	dumpdimensions(node);
	fprintf(stdout," &%lx",(unsigned long)node);
	fprintf(stdout,"\n");
	fprintf(stdout,"%sarray:\n",dent2(depth+1));
	dumpocnode1((OCnode*)oclistget(node->subnodes,0),depth+2);
	fprintf(stdout,"%smaps:\n",dent2(depth+1));
	for(i=1;i<oclistlength(node->subnodes);i++) {
	    dumpocnode1((OCnode*)oclistget(node->subnodes,i),depth+2);
	}
    } break;

    case OC_Attribute: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	if(node->name == NULL) OCPANIC("Attribute without name");
	fprintf(stdout,"%s %s",octypetostring(node->basetype),node->name);
	for(n=0;n<oclistlength(node->att.values);n++) {
	    char* value = (char*)oclistget(node->att.values,n);
	    if(n > 0) fprintf(stdout,",");
	    fprintf(stdout," %s",value);
	}
	fprintf(stdout," &%lx",(unsigned long)node);
	fprintf(stdout,"\n");
    } break;

    case OC_Attributeset: {
        fprintf(stdout,"[%2d]%s ",depth,dent(depth));
	fprintf(stdout,"%s:\n",node->name?node->name:"Attributes");
	for(n=0;n<oclistlength(node->subnodes);n++) {
	    dumpocnode1((OCnode*)oclistget(node->subnodes,n),depth+1);
	}
    } break;

    default:
	OCPANIC1("encountered unexpected node type: %x",node->octype);
    }

    if(node->attributes != NULL) {
	unsigned int i;
	for(i=0;i<oclistlength(node->attributes);i++) {
	    OCnode* att = (OCnode*)oclistget(node->attributes,i);
	    OCattinfo* info = &att->att;
	    size_t nvalues;
	    fprintf(stdout,"%s[%s=",dent2(depth+2),att->name);
	    nvalues = oclistlength(info->values);
	    if(nvalues == 0)
		OCPANIC("|Attributes| == 0");
	    if(nvalues == 1) {
		dumpattvalue(att->basetype,(char*)oclistget(info->values,0));
	    } else {
		unsigned int j;
	        fprintf(stdout,"{");
		for(j=0;j<nvalues;j++) {
		    if(j>0) fprintf(stdout,", ");
		    dumpattvalue(att->basetype,(char*)oclistget(info->values,j));
		}
	        fprintf(stdout,"}");
	    }
	    fprintf(stdout,"]\n");
	}
    }
}

static void
dumpdimensions(OCnode* node)
{
    unsigned int i;
    for(i=0;i<node->array.rank;i++) {
        OCnode* dim = (OCnode*)oclistget(node->array.dimensions,i);
        fprintf(stdout,"[%s=%lu]",
			(dim->name?dim->name:"?"),
			(unsigned long)dim->dim.declsize);
    }
}

static void
dumpattvalue(OCtype nctype, const char* string)
{
    if(nctype == OC_String || nctype == OC_URL) {
        fprintf(stdout,"\"%s\"",string);
    } else {
        fprintf(stdout,"%s",string);
    }
}

void
ocdumpslice(OCslice* slice)
{
    fprintf(stdout,"[");
    fprintf(stdout,"%lu",(unsigned long)slice->first);
    if(slice->stride > 1) fprintf(stdout,":%lu",(unsigned long)slice->stride);
    fprintf(stdout,":%lu",(unsigned long)(slice->first+slice->count)-1);
    fprintf(stdout,"]");
}

void
ocdumpclause(OCprojectionclause* ref)
{
    unsigned int i;
    OClist* path = oclistnew();
    occollectpathtonode(ref->node,path);
    for(i=0;i<oclistlength(path);i++) {
        OClist* sliceset;
	OCnode* node = (OCnode*)oclistget(path,i);
	if(node->tree != NULL) continue; /* leave off the root node*/
	fprintf(stdout,"%s%s",(i>0?PATHSEPARATOR:""),node->name);
	sliceset = (OClist*)oclistget(ref->indexsets,i);
	if(sliceset != NULL) {
	    unsigned int j;
	    for(j=0;j<oclistlength(sliceset);j++) {
	        OCslice* slice = (OCslice*)oclistget(sliceset,j);
	        ocdumpslice(slice);
	    }
	}
    }
}


static void
addfield(char* field, char* line, int align)
{
    int len,rem;
    strcat(line,"|");
    strcat(line,field);
    len = strlen(field);
    rem = (align - len);
    while(rem-- > 0) strcat(line," ");
}

static void
dumpfield(int index, char* n8, int isxdr)
{
    char line[1024];
    char tmp[32];

    union {
	unsigned int uv;
	int sv;
	char cv[4];
	float fv;
    } form;
    union {
	char cv[8];
        unsigned long long ll;
        double d;
    } dform;

    line[0] = '\0';

    /* offset */
    sprintf(tmp,"%6zd",index);
    addfield(tmp,line,5);

    memcpy(form.cv,n8,4);

    /* straight hex*/
    sprintf(tmp,"%08x",form.uv);
    addfield(tmp,line,8);

    if(isxdr) {swapinline32(&form.uv);}

    /* unsigned integer */
    sprintf(tmp,"%12u",form.uv);
    addfield(tmp,line,12);

    /* signed integer */
    sprintf(tmp,"%12d",form.sv);
    addfield(tmp,line,12);

    /* float */
    sprintf(tmp,"%#g",form.fv);
    addfield(tmp,line,12);

    /* char[4] */
    {
        /* use raw form (i.e. n8)*/
        int i;
	tmp[0] = '\0';
        for(i=0;i<4;i++) {
	    char stmp[64];
	    unsigned int c = (n8[i] & 0xff);
	    if(c < ' ' || c > 126)
                snprintf(stmp,sizeof(stmp),"\\%02x",c);
	    else
                snprintf(stmp,sizeof(stmp),"%c",c);
	    if(!occoncat(tmp,sizeof(tmp),1,stmp))
		return;
        }
    }

    addfield(tmp,line,16);

    /* double */
    memcpy(dform.cv,n8,(size_t)(2*XDRUNIT));
    if(isxdr) xxdrntohdouble(dform.cv,&dform.d);
    sprintf(tmp,"%#g",dform.d);
    addfield(tmp,line,12);

    fprintf(stdout,"%s\n",line);
}

static void
typedmemorydump(char* memory, size_t len, int fromxdr)
{
    unsigned int i,count,rem;
    char line[1024];
    char* pmem;
    char mem[8];

    assert(memory[len] == 0);

    /* build the header*/
    line[0] = '\0';
    addfield("offset",line,6);
    addfield("hex",line,8);
    addfield("uint",line,12);
    addfield("int",line,12);
    addfield("float",line,12);
    addfield("char[4]",line,16);
    addfield("double",line,12);
    strcat(line,"\n");
    fprintf(stdout,"%s",line);

    count = (len / sizeof(int));
    rem = (len % sizeof(int));

    for(pmem=memory,i=0;i<count;i++,pmem+=4) {
	memset(mem,0,8);
	if(i<(count-1))
	    memcpy(mem,pmem,8);
	else
	    memcpy(mem,pmem,4);
	dumpfield((int)(i*sizeof(size_t)),mem,fromxdr);
    }
    if(rem > 0) {
	memset(mem,0,8);
	memcpy(mem,pmem,4);
	dumpfield((int)(i*sizeof(size_t)),mem,fromxdr);
    }
    fflush(stdout);
}

static void
simplememorydump(char* memory, size_t len, int fromxdr)
{
    unsigned int i,count,rem;
    int* imemory;
    char tmp[32];
    char line[1024];

    assert(memory[len] == 0);

    /* build the header*/
    line[0] = '\0';
    addfield("offset",line,6);
    addfield("XDR (hex)",line,9);
    addfield("!XDR (hex)",line,10);
    fprintf(stdout,"%s\n",line);

    count = (len / sizeof(int));
    rem = (len % sizeof(int));
    if(rem != 0)
	fprintf(stderr,"ocdump: |mem|%%4 != 0\n");
    imemory = (int*)memory;

    for(i=0;i<count;i++) {
	unsigned int vx = (unsigned int)imemory[i];
	unsigned int v = vx;
	if(!xxdr_network_order) swapinline32(&v);
        line[0] = '\0';
        sprintf(tmp,"%6d",i);
        addfield(tmp,line,6);
        sprintf(tmp,"%08x",vx);
        addfield(tmp,line,9);
        sprintf(tmp,"%08x",v);
        addfield(tmp,line,10);
        fprintf(stdout,"%s\n",line);
    }
    fflush(stdout);
}

void
ocdumpmemory(char* memory, size_t len, int xdrencoded, int level)
{
    if(level > MAXLEVEL) level = MAXLEVEL;
    switch (level) {
    case 1: /* Do a multi-type dump */
        typedmemorydump(memory,len,xdrencoded);
	break;
    case 0: /* Dump a simple linear list of the contents of the memory as 32-bit hex and decimal */
    default:
        simplememorydump(memory,len,xdrencoded);
	break;
    }
}

static int
ocreadfile(FILE* file, size_t datastart, char** memp, size_t* lenp)
{
    char* mem;
    size_t len;
    long red;
    struct stat stats;
    long pos;

    pos = ftell(file);
    fseek(file,0,SEEK_SET);
    fseek(file,(long)datastart,SEEK_SET);

    fstat(fileno(file),&stats);
    len = stats.st_size;
    len -= datastart;
    
    mem = (char*)calloc(len+1,1);
    if(mem == NULL) return 0;

    /* Read only the data part */
    red = fread(mem,1,len,file);
    if(red < len) {
	fprintf(stderr,"ocreadfile: short file\n");
	return 0;
    }	
    fseek(file,pos,SEEK_SET); /* leave it as we found it*/
    if(memp) *memp = mem;
    if(lenp) *lenp = len;
    return 1;
}

void
ocdd(OCstate* state, OCnode* root, int xdrencoded, int level)
{
    char* mem;
    size_t len;
    if(root->tree->data.file != NULL) {
        if(!ocreadfile(root->tree->data.file,root->tree->data.bod,&mem,&len)) {
    	    fprintf(stderr,"ocdd could not read data file\n");
	    return;
	}
        ocdumpmemory(mem,len,xdrencoded,level);
        free(mem);
    } else {
        mem = root->tree->data.memory;
        mem += root->tree->data.bod;
        len = root->tree->data.datasize;
        len -= root->tree->data.bod;
        ocdumpmemory(mem,len,xdrencoded,level);
    }
}

void
ocdumpdata(OCstate* state, OCdata* data, OCbytes* buffer, int frominstance)
{
    char tmp[1024];
    OCnode* template = data->template;
    snprintf(tmp,sizeof(tmp),"%lx:",(unsigned long)data);
    ocbytescat(buffer,tmp);
    if(!frominstance) {
        ocbytescat(buffer," node=");
        ocbytescat(buffer,template->name);
    }
    snprintf(tmp,sizeof(tmp)," xdroffset=%ld",(unsigned long)data->xdroffset);
    ocbytescat(buffer,tmp);
    if(data->template->octype == OC_Atomic) {
        snprintf(tmp,sizeof(tmp)," xdrsize=%ld",(unsigned long)data->xdrsize);
        ocbytescat(buffer,tmp);
    }
    if(iscontainer(template->octype)) {
        snprintf(tmp,sizeof(tmp)," ninstances=%d",(int)data->ninstances);
        ocbytescat(buffer,tmp);
    } else if(template->basetype == OC_String || template->basetype == OC_URL) {
        snprintf(tmp,sizeof(tmp)," nstrings=%d",(int)data->nstrings);
        ocbytescat(buffer,tmp);
    }
    ocbytescat(buffer," container=");
    snprintf(tmp,sizeof(tmp),"%lx",(unsigned long)data->container);
    ocbytescat(buffer,tmp);
    ocbytescat(buffer," mode=");
    ocbytescat(buffer,ocdtmodestring(data->datamode,0));
}

/*
Depth Offset   Index Flags Size Type      Name
0123456789012345678901234567890123456789012345
0         1         2         3         4
[001] 00000000 0000  FS    0000 Structure person
*/
static const int tabstops[] = {0,6,15,21,27,32,42};
static const char* header =
"Depth Offset   Index Flags Size Type     Name\n";

void
ocdumpdatatree(OCstate* state, OCdata* data, OCbytes* buffer, int depth)
{
    size_t i,rank;
    OCnode* template;
    char tmp[1024];
    size_t crossproduct;
    int tabstop = 0;
    const char* typename;

    /* If this is the first call, then dump a header line */
    if(depth == 0) {
	ocbytescat(buffer,header);
    }

    /* get info about the template */
    template = data->template;
    rank = template->array.rank;

    /* get total dimension size */
    if(rank > 0)
        crossproduct = octotaldimsize(template->array.rank,template->array.sizes);

    /* Dump the depth first */
    snprintf(tmp,sizeof(tmp),"[%03d]",depth);
    ocbytescat(buffer,tmp);

    tabto(tabstops[++tabstop],buffer);

    snprintf(tmp,sizeof(tmp),"%08lu",(unsigned long)data->xdroffset);
    ocbytescat(buffer,tmp);

    tabto(tabstops[++tabstop],buffer);

    /* Dump the Index wrt to parent, if defined */
    if(fisset(data->datamode,OCDT_FIELD)
       || fisset(data->datamode,OCDT_ELEMENT)
       || fisset(data->datamode,OCDT_RECORD)) {
        snprintf(tmp,sizeof(tmp),"%04lu ",(unsigned long)data->index);
        ocbytescat(buffer,tmp);
    }
   
    tabto(tabstops[++tabstop],buffer);

    /* Dump the mode flags in compact form */
    ocbytescat(buffer,ocdtmodestring(data->datamode,1));
   
    tabto(tabstops[++tabstop],buffer);

    /* Dump the size or ninstances */
    if(fisset(data->datamode,OCDT_ARRAY)
       || fisset(data->datamode,OCDT_SEQUENCE)) {
        snprintf(tmp,sizeof(tmp),"%04lu",(unsigned long)data->ninstances);
    } else {
        snprintf(tmp,sizeof(tmp),"%04lu",(unsigned long)data->xdrsize);
    }
    ocbytescat(buffer,tmp);

    tabto(tabstops[++tabstop],buffer);

    if(template->octype == OC_Atomic) {
	typename = octypetostring(template->basetype);
    } else { /*must be container*/
	typename = octypetostring(template->octype);
    }
    ocbytescat(buffer,typename);

    tabto(tabstops[++tabstop],buffer);

    if(!occopycat(tmp,sizeof(tmp),1,template->name))
	return;
    ocbytescat(buffer,tmp);

    if(rank > 0) {
	snprintf(tmp,sizeof(tmp),"[%lu]",(unsigned long)crossproduct);
	ocbytescat(buffer,tmp);
    }
    ocbytescat(buffer,"\n");

    /* dump the sub-instance, which might be fields, records, or elements */
    if(!fisset(data->datamode,OCDT_ATOMIC)) {
        for(i=0;i<data->ninstances;i++)
	    ocdumpdatatree(state,data->instances[i],buffer,depth+1);
    }
}

void
ocdumpdatapath(OCstate* state, OCdata* data, OCbytes* buffer)
{
    int i;
    OCdata* path[1024];
    char tmp[1024];
    OCdata* pathdata;
    OCnode* template;
    int isrecord;

    path[0] = data;
    for(i=1;;i++) {
	OCdata* next = path[i-1];
	if(next->container == NULL) break;
	path[i] = next->container;
    }    	        
    /* Path is in reverse order */
    for(i=i-1;i>=0;i--) {
	pathdata = path[i];
	template = pathdata->template;
	ocbytescat(buffer,"/");
	ocbytescat(buffer,template->name);
	/* Check the mode of the next step in path */
	if(i > 0) {
	    OCdata* next = path[i-1];
	    if(fisset(next->datamode,OCDT_FIELD)
		|| fisset(next->datamode,OCDT_ELEMENT)
		|| fisset(next->datamode,OCDT_RECORD)) {
		snprintf(tmp,sizeof(tmp),".%lu",(unsigned long)next->index);
	        ocbytescat(buffer,tmp);
	    }
	}
	if(template->octype == OC_Atomic) {
	    if(template->array.rank > 0) {
	        size_t xproduct = octotaldimsize(template->array.rank,template->array.sizes);
	        snprintf(tmp,sizeof(tmp),"[0..%lu]",(unsigned long)xproduct-1);
	        ocbytescat(buffer,tmp);
	    }
	}
	isrecord = 0;
	if(template->octype == OC_Sequence) {
	    /* Is this a record or a sequence ? */
	    isrecord = (fisset(pathdata->datamode,OCDT_RECORD) ? 1 : 0);
	}
    }
    /* Add suffix to path */
    if(iscontainer(template->octype)) {
        /* add the container type, except distinguish record and sequence */
	ocbytescat(buffer,":");
	if(isrecord)
	    ocbytescat(buffer,"Record");
	else
	    ocbytescat(buffer,octypetostring(template->octype));
    } else if(isatomic(template->octype)) {
	/* add the atomic basetype */
	ocbytescat(buffer,":");
	ocbytescat(buffer,octypetostring(template->basetype));
    }
    snprintf(tmp,sizeof(tmp),"->0x%0lx",(unsigned long)pathdata);
    ocbytescat(buffer,tmp);	
}
